ada_feeding_perception/ada_feeding_perception/segment_all_items.py

```diff
@@ -0,0 +1,1279 @@
+        boxes = outputs["pred_boxes"][0]
+        self._node.get_logger().info("... Done")
+
+        # Filter the output based on the box and text thresholds
+        boxes_cxcywh = {}
+        logits_filt = logits.cpu().clone()
+        boxes_filt = boxes.cpu().clone()
+        filt_thresh_mask = logits_filt.max(dim=1)[0] > box_threshold
+        logits_filt = logits_filt[filt_thresh_mask]
+        boxes_filt = boxes_filt[filt_thresh_mask]
+
+        # Tokenize the caption
+        tokenizer = self.groundingdino.tokenizer
+        caption_tokens = tokenizer(caption)
+
+        # Build the dictionary of bounding boxes for each food item label detected
+        for logit, box in zip(logits_filt, boxes_filt):
+            # Predict phrases based on the bounding boxes and the text threshold
+            phrase = get_phrases_from_posmap(
+                logit > text_threshold, caption_tokens, tokenizer
+            )
+            if phrase not in boxes_cxcywh:
+                boxes_cxcywh[phrase] = []
+            boxes_cxcywh[phrase].append(box.cpu().numpy())
+
+        # Define height and width of image
+        height, width, _ = image.shape
+
+        # Convert the bounding boxes outputted by GroundingDINO to the following format
+        # [top left x-value, top left y-value, bottom right x-value, bottom right y-value]
+        # and unnormalize the bounding box coordinate values
+        boxes_xyxy = {}
+        for phrase, boxes in boxes_cxcywh.items():
+            boxes_xyxy[phrase] = []
+            for box in boxes:
+                # Scale the box from percentage values to pixel values
+                box = np.multiply(box, np.array([width, height, width, height]))
+                center_x, center_y, w, h = box
+                # Get the bottom left and top right coordinates of the box
+                x0 = center_x - (w / 2)
+                y0 = center_y - (h / 2)
+                x1 = x0 + w
+                y1 = y0 + h
+                boxes_xyxy[phrase].append([x0, y0, x1, y1])
+
+        # Measure the elapsed time running GroundingDINO on the image prompt
+        inference_time = int(round((time.time() - inference_time) * 1000))
+
+        return boxes_xyxy
+
+    def load_image(self, image_array: npt.NDArray):
+        """
+        Load the image and apply transformations to it.
+
+        Parameters
+        ----------
+        image_array: The image to load.
+
+        Returns
+        -------
+        image_pil: The image in Image pillow format.
+        image: The image in tensor format.
+        """
+        # Convert image from BGR to RGB to convert CV2 image to image pillow
+        image_array = cv2.cvtColor(image_array, cv2.COLOR_BGR2RGB)
+
+        # Convert image to image pillow to apply transformation
+        image_pil = ImagePIL.fromarray(image_array, mode="RGB")
+        transform = T.Compose(
+            [
+                T.RandomResize([800], max_size=1333),
```

Comment:
What all transforms do you do? To make this function more general, I'd suggest you make them parameters, with default values. (i.e., it looks like you're resizing the image to size 800x800, but maybe in the future we may want to resize the image to a different size).